// @chub-ai/stages-ts/dist/types/defaults.js
export const DEFAULT_INITIAL = {
    environment: 'development',
    initState: null,
    characters: {}, config: null, messageState: null, users: {}, chatState: null
};
export const DEFAULT_CHARACTER = {
    alternate_greetings: [],
    anonymizedId: "1",
    description: "",
    example_dialogs: "",
    first_message: "",
    isRemoved: false,
    name: "",
    partial_extensions: {},
    personality: "",
    post_history_instructions: "",
    scenario: "",
    system_prompt: "",
    tavern_personality: ""
};
export const DEFAULT_USER = {
    anonymizedId: "0", chatProfile: "", isRemoved: false, name: ""
};
export const DEFAULT_MESSAGE = {
    anonymizedId: "0", content: "", isBot: false, promptForId: "1",
    identity: '12345', isMain: true
};
export const DEFAULT_LOAD_RESPONSE = {
    error: null, success: true, chatState: null, initState: null, messageState: null
};
export const DEFAULT_RESPONSE = {
    chatState: null, systemMessage: null,
    error: null, extensionMessage: null, modifiedMessage: null,
    messageState: null, stageDirections: null
};
