// @chub-ai/stages-ts/dist/types/generation/defaults.js
import { AspectRatio } from "./images";
import { CHUBBY_KITTY } from "./constants";
export const DEFAULT_GENERATION_REQUEST = {
    seed: null, item_id: null
};
export const DEFAULT_PROMPTED_REQUEST = {
    negative_prompt: null,
    prompt: "Super cute chubby kitten, high res perfect, will make you believe in the afterlife, 8K",
};
export const DEFAULT_IMAGE_REQUEST = Object.assign(Object.assign({ aspect_ratio: AspectRatio.SQUARE, remove_background: false }, DEFAULT_PROMPTED_REQUEST), DEFAULT_GENERATION_REQUEST);
export const DEFAULT_IMAGE_TO_IMAGE_REQUEST = Object.assign(Object.assign({}, DEFAULT_IMAGE_REQUEST), { image: CHUBBY_KITTY, strength: 0.7 });
export const DEFAULT_ANIMATION_REQUEST = Object.assign(Object.assign({}, DEFAULT_GENERATION_REQUEST), { image: CHUBBY_KITTY, cfg_scale: 1.8, motion_bucket_id: 127 });
export const DEFAULT_INPAINT_REQUEST = Object.assign(Object.assign({}, DEFAULT_IMAGE_TO_IMAGE_REQUEST), { search_prompt: "Anything with a face", mask: null });
export const DEFAULT_BACKGROUND_REMOVAL_REQUEST = {
    image: CHUBBY_KITTY
};
export const DEFAULT_DURATION = {
    seconds: 5,
};
export const DEFAULT_TEXT_TO_VIDEO_REQUEST = Object.assign(Object.assign({}, DEFAULT_PROMPTED_REQUEST), DEFAULT_DURATION);
export const DEFAULT_FOLEY_REQUEST = Object.assign(Object.assign(Object.assign({}, DEFAULT_DURATION), DEFAULT_GENERATION_REQUEST), { prompt: 'Thunderstorm' });
export const DEFAULT_TTS_REQUEST = {
    transcript: "The quick brown fox jumps over the lazy dog.", voice_id: null
};
export const DEFAULT_MUSIC_REQUEST = {
    prompt: "Elevator music with dementia",
    lyrics: null, tags: null, title: null, instrumental: false,
    lyrics_prompt: null
};
export const DEFAULT_TEXT_PROMPT = {
    prompt: '', stop: [], max_tokens: 50, include_history: false, template: '', context_length: null,
    min_tokens: null
};
